### ISSUE TITLE
Literal type branching flow

### ISSUE THREAD

**Issue by tamuhey** [microsoft/pyright#1437]
I want pyright to infer the below code is no problem:

```python
A = Literal["a", "b"]


def f(a: A) -> int:
    if a == "a":
        return 0
    elif a == "b": # I don't want to change `elif` to `else`, to prevent a bug when `A` has more literal value.
        return 1
    # Here is unreachable, but..
    # ERROR: Type "None" cannot be assigned to type "int"
```

Is it technically possible?

**Comment by izzudinhafiz** [comment]
`# Here is unreachable, but..`

That section is _not_ unreachable. If `a` is not "a" or "b", you will reach that line and the function will return None.

I assume you say its unreachable because your code will (hopefully) never pass anything other than "a" or "b" but pyright cannot know that.

The solution I think is to raise a `ValueError` which will remove that error you mentioned.

```python
A = Literal["a", "b"]

def f(a: A) -> int:
    if a == "a":
        return 0
    elif a == "b": # I don't want to change `elif` to `else`, to prevent a bug when `A` has more literal value.
        return 1
    raise ValueError("function paramater must be 'a' or 'b'")
```

**Comment by erictraut** [comment]
You have annotated parameter `a` in a way that any attempt to pass a value other than `"a"` or `"b"` will result in a type violation, so why are you reluctant to replace the `elif` with an `else`?

**Comment by tamuhey** [comment]
@erictraut Because if I add new value to type `A`, for example, `A = Literal["a", "b", "c"]`, there may be a bug in that code. The function `f` should be modified to have `elif a == "c"`, but I may miss it:

```python
A = Literal["a", "b", "c"] # <= add "c" to the literal `A`


def f(a: A) -> int:
    if a == "a":
        return 0
    else:   # <= I should modify this line to process `a=="c"`. But there is no chance for any static analyzer to suggest me to modify.
        return 1
```

However, if I use `elif a == "b"`, Pyright can detect there is a problem:

```python
A = Literal["a", "b", "c"] # <= add "c" to the literal `A`


def f(a: A) -> int:
    if a == "a":
        return 0
    elif a == "b":
        return 1
    # Error: "c" is not handled.
```

Of course, as @izzudinhafiz suggested, there is a way to solve this problem, adding `raise` at the end of `f`.

**Comment by erictraut** [comment]
OK, that's legitimate. I'll look into it.

**Comment by layday** [comment]
Apropos of nothing, I remember reading about this trick on a blog somewhere many moons ago, which works to catch omission errors even when the conditional isn't coupled to the return type:

```py
from typing import Literal, NoReturn, Union

A = Literal["a", "b"]


def test_never(
    value: Union[()],  # Empty union is of type Never
) -> NoReturn:
    raise ValueError("received", value)


def foo(a: Literal["a", "b"]) -> int:
    if a == "a":
        return 0
    elif a == "b":
        return 1
    # "a" is of type Never; no error
    test_never(a)


def bar(a: Literal["a", "b", "c"]) -> None:
    if a == "a":
        ...
    elif a == "b":
        ...
    else:
        # Argument of type "Literal['c']" cannot be assigned to parameter "value" of type "Never" in function "test_never"
        #   Type "Literal['c']" cannot be assigned to type "Never"
        test_never(a)
```

**Comment by erictraut** [comment]
Yes, thanks for the suggestion @layday. The expression `Union[()]` isn't legal, but you can use `NoReturn`, which has the semantics of an empty union. This use of `NoReturn` is recommended by the mypy team, and it works with Pyright as well.

**Comment by tamuhey** [comment]
I think @layday 's code is a solution for this issue - it prevents both of runtime error and type error.

**Comment by erictraut** [comment]
I've explored what it would take to determine whether a code flow location is reachable based on type narrowing. It's technically possible, but it will have a significant performance impact on type analysis, especially in cases where the function's return type is not annotated and must be inferred from the implementation. For this reason, I'm very reluctant to add the functionality. It would be a net negative to most users.

I also looked at the TypeScript compiler and confirmed that it also doesn't do this.

I think that the suggestion above is the best way to address this problem.
